src/aurora/mod.rs

```diff
@@ -377,8 +377,8 @@ where
         }
 
         // Resize the instance to the padded length
-        let mut instance = instance;
-        instance.resize(num_instance_variables, F::ZERO);
+        let mut zero_padded_instance = instance.clone();
```

Comment:
But if the caller will no longer be needing the instance, this cloning we are doing here wouldn't be necessary, right? (which is how it was before.) That's why transferring ownership and deferring the decision on whether to clone or not to the caller (who knows whether the data in the proof is needed later on or not) is generally a good idea. Also note that the `*`s introduced below simply call `clone()` (as far as I know), which is only needed because things are passed as references now.

These are likely all minor costs, but I do not quite see the benefit in the switch to passing references.

Reply:
@Antonio95 notice that we are anyway cloning below for the second resize.

Reply:
I agree that if we were NOT cloning before and we are now, then that's not a good design.
However, I think this underlines a larger design issue: I don't think the verifier should ever *have to* clone nor take ownership of things like proof or instance.
Regarding the instance:
Think of the case where a single verifier has multiple proofs for the same instance. Instance should be passable by reference.

Reply:
Think about the following: the verifier will need to use the actual values (say, `F`s) contained in the vectors of the proof. If you pass something by reference, these have to be cloned - one just doesn't see it because our small types (such as `F`) implement `Copy` and therefore `clone()` gets called automatically (e.g. when one dereferences them: `*v`). But this isn't slower (but instead possibly faster than) than calling `vector.clone()` (where `vector` is the `Vec` that contains those `v`s). To be fair, the latter is more costly memory-wise, since one needs to have the entire vector in memory at once instead of cloning its elements one by one.

Regarding the case you mentioned, where a single verifier has multiple proofs for the same instance, that is exactly why I advocate for passing things by ownership and deferring the decision of whether to clone or not to the caller - they indeed know whether the instance in question is necessary later on or not. If one passes things by reference, the callee has to pay the cloning cost in cases where it wouldn't be necessary (e.g. instance used only in one proof).

Is what I'm trying to say clear? There's sth I don't like about my own argument, which is that it seems to lead to the conclusion that one should _always_ move all arguments instead of passing references - which is not what we see in Rust references and a lot of good Rust code. Perhaps the only situation where one way or another it seems clear that passing by reference is a good choice is when the structure is very large, since then one likely doesn't want to copy it onto the stack (where all function arguments go). However, I don't think this is a very common situation. Allocated types such as `Vec` live in the heap and the structure to which they belong only contains a pointer to them.

Reply:
> the verifier will need to use the actual values (say, Fs) contained in the vectors of the proof. If you pass something by reference, these have to be cloned

I believe this is not correct, but I'm not an expert here so please take this with a grain of salt.

`Vec` doesn't implement `Copy`, you always have to `Clone` it, which is expensive (on the heap). On the other, you are right that eventually the underlying representation needs to be copied (!= cloned). In our case, the verifier actually doesn't use the elements `F` directly,  he only needs references to it for the `inner_product`.  

Under the hood, the `inner_product` multiplies two `&F`s together. Further under the hood, this is done by [accessing the `u64` stored there](https://github.com/arkworks-rs/algebra/blob/30ad93b079d345bef03f53d92b1ef7017936af85/ff/src/fields/models/fp/montgomery_backend.rs#L188) - at which point a `Copy` occurs. IIUC, this is done on the stack already, and is therefore very cheap.

So, the bottom line is that cloning `Vec` explicitly is more expensive than passing it by reference and only copying the underlying value onto the stack when iterating through it.

Reply:
> There's sth I don't like about my own argument, which is that it seems to lead to the conclusion that one should always move all arguments instead of passing references - which is not what we see in Rust references and a lot of good Rust code. 

This really comes down to the types that we're passing around. If these types implement `Copy`, then usually this means it's cheap to pass them "by value" and the compiler will take care of cloning if it encounters the need in the caller.
Since `Vec` doesn't have `Copy` and cloning is expensive, it's not good practice to pass by `Vec`s by value and clone all the time.

Reply:
Cf. long message below. Good job @mmagician !